File: versions/forge1710/src/common/java/com/idkidknow/mcreallink/forge1710/events/ModCommand.java
package com.idkidknow.mcreallink.forge1710.events;

import net.minecraft.command.CommandBase;
import net.minecraft.command.CommandException;
import net.minecraft.command.ICommand;
import net.minecraft.command.ICommandSender;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Optional;
import java.util.function.Supplier;

public class ModCommand extends CommandBase {
    private ModCommand() {}
    private static final ModCommand instance = new ModCommand();
    public static ModCommand getInstance() { return instance; }

    private Supplier<Optional<Throwable>> startAction = Optional::empty;
    public void setStartAction(Supplier<Optional<Throwable>> action) {
        this.startAction = action;
    }
    private Runnable stopAction = () -> {};
    public void setStopAction(Runnable action) {
        this.stopAction = action;
    }
    private Runnable downloadAction = () -> {};
    public void setDownloadAction(Runnable action) {
        this.downloadAction = action;
    }

    @Override
    public String getCommandName() {
        return "reallink";
    }

    @Override
    public String getCommandUsage(ICommandSender sender) {
        return "/reallink start\n/reallink stop\n/reallink download";
    }

    @Override
    public void processCommand(ICommandSender iCommandSender, String[] args) {
        if (args.length == 0) return;
        if (args[0].equals("start")) {
            Optional<Throwable> ret = startAction.get();
            if (ret.isPresent()) {
                throw new CommandException("Failed", ret.get());
            }
        } else if (args[0].equals("stop")) {
            stopAction.run();
        } else if (args[0].equals("download")) {
            downloadAction.run();
        }
    }

    @Override
    public List<String> addTabCompletionOptions(ICommandSender sender, String[] args) {
        return args.length >= 1 ? getListOfStringsMatchingLastWord(args, "start", "stop", "download") : null;
    }

    @Override
    public int compareTo(@NotNull Object o) {
        return this.getCommandName().compareTo(((ICommand) o).getCommandName());
    }
}

File: forge1_7_10/src/com/idkidknow/realitylink/platform/ModCommand.scala
package com.idkidknow.realitylink.platform

import net.minecraft.command.CommandBase
import net.minecraft.command.CommandException
import net.minecraft.command.ICommandSender

object ModCommand extends CommandBase {
  // scalafix:off DisableSyntax.var
  var startAction: () => Either[Throwable, Unit] = () => Right(())
  var stopAction: () => Unit = () => {}
  var downloadAction: () => Unit = () => {}
  // scalafix:on
  override def getCommandName: String = "realitylink"
  override def getCommandUsage(player: ICommandSender): String =
    "/realitylink start\n/realitylink stop\n/realitylink download"
  @SuppressWarnings(Array("scalafix:DisableSyntax.throw"))
  override def processCommand(
      sender: ICommandSender,
      args: Array[String],
  ): Unit = {
    args match {
      case Array("start") =>
        startAction() match {
          case Left(e) => throw CommandException("Failed to start", e)
          case _ =>
        }
      case Array("stop") =>
        stopAction()
      case Array("download") =>
        downloadAction()
      case _ => throw CommandException("Invalid arguments")
    }
  }
  override def addTabCompletionOptions(
      sender: ICommandSender,
      args: Array[String],
  ): java.util.List[String] = {
    if (args.length == 1) // scalafix:ok
      CommandBase
        .getListOfStringsMatchingLastWord(args, "start", "stop", "download")
        .asInstanceOf // scalafix:ok
    else java.util.Collections.emptyList
  }
}
